### ISSUE TITLE
Blazor components in librarries with code-behind, and a modular structure not recognizing the defined components

### ISSUE THREAD

**Issue by KKacer** [dotnet/aspnetcore#19263]
Found some issues earlier with the '.razor' files not recognizing my comonents in a modular platform.

`https://postimg.cc/SXC0Rx04`

Currently I simplified the structure as possible with just a simple example showing the issue.
Always experiencing "markup element with unexpected name BlahBlah". 

[https://github.com/KKacer/RecreateBlazorError](url)

**Update:** _In the last commit moved the code-behind to the .razor files, issues still persist._

Tried to use Code-behind for the razor files. Which seems can't make them linked in the component library side, which may be cause of some components being unrecognized.

The project is just the bare-bone  and minimum structure of my codes, I minimized all the modules and complicated folder structure to the most simple forms. 
Maybe I made a mistake, or it needs a hack currently, but mostly seems like a bug. As I remember Visual Studio and Asp.net always seemed buggy in this aspect when dealing with a more complex folder and projects structure from my side and what I experienced.

**Update:**
Aside from the code-behind, tested the same also without code-behind, the issues persist.
See a new issue also there inside this recreation project which is related to finding namespaces, even after a rebuild.

`https://postimg.cc/hX5gft2Z/2dc81580`

Please care more and provide more support for more custom and complex folder, project and namespaces structure.

**Comment by chrissainty** [comment]
Hi @KKacer,

I've had a look at your solution and I think you're having issues because you've put your components in a .NET Standard class library and not a Razor Class Library. Blazor components can only be shared via Razor Class Libraries, might be worth [checking out that section of the docs](https://docs.microsoft.com/en-gb/aspnet/core/blazor/class-libraries?view=aspnetcore-3.1&tabs=visual-studio) if you've not had a chance already.

**Comment by KKacer** [comment]
@chrissainty 
Nice point, earlier was building components with "Razor Class Library", but I already have some other cloned projects which they also used "Standard libraries" for that purpose, in addition, I used exactly the same Nuget packages, there were nothing else.
Cause my main concerns and goals are re-usability and modularity, thought it could be ideal to use .Net Standard libraries, am I wrong?

**Comment by chrissainty** [comment]
Sorry, I don’t quite understand everything you’ve said. But, if you want to share/reuse Blazor components you must use Razor Class Libraries.

**Comment by KKacer** [comment]
Got that @chrissainty 
Made this post before our last discussion, and before the last commit.
Sorry if there were any grammatical issues here, bypassed the issue via what you said.
In earlier post just mentioned that, even when using "Razor Class Library" also had some issues making ".razor" file understand the components.

In the latest build a namespace was unknown, which will post the image shot below:

`https://postimg.cc/RWyB0vmT/a3514bb7`

**Comment by KKacer** [comment]
I have a working version of this sample, but still trying to figure it out that why several similar errors happen, some randomly.

I know that we can build it finally by some hacks or extra effort.
But don't understand this, while this is only the simplest form, why should we encounter any hassle on this case.
Just anyone interested please have a look at this video:
https://1drv.ms/v/s!AlScPmE9PLAKgRJbuHx0mFPHFMY7?e=IbV5Z2
the last build on git:
https://github.com/KKacer/RecreateBlazorError

I have another working version, but can't see any difference, and don't know why such problems happen randomly. while you made a project from the ground up.
At least the environment could provide some helpful build errors, if there were anything really wrong.

**Comment by javiercn** [comment]
@KKacer Thanks for the details.

We'll look into this and get back to you.

@pranavkm the error on the video seems to point to some error in the application parts discovery logic.

**Comment by KKacer** [comment]
Welcome, 
If any of the images gone by the time you need, let me know to re-upload.
I also try to be available for more details and discussions when needed.
and thanks for the effort and care you put on it.

**Comment by KKacer** [comment]
Another similar issue to the first described problem which was discussed earlier:

https://1drv.ms/v/s!AlScPmE9PLAKgRPsNqW-qLmi-_2h?e=OejKK8

The project was working fine, flawlessly, just committed the new changes to Azure DevOps, and right after that again it was saying that it can't understand some components there!
Earlier also such a thing was happening, don't know how it went away, guess just quit and reopened VS and it got fixed, but believe me these are annoying, and I should not rebuild every time, though even rebuilding projects some times doesn't fix these issues.
To clarify more, there is no namespace or referencing issues there, and the project was running as expected.

My comment is when you try modular development with multiple assemblies, custom folder structure and custom namespace rules targeting more reusability across your projects, these types of errors happen more, by renaming a component, and assembly, and taking other actions they may appear and disappear.

**Comment by KKacer** [comment]
Another bug:

https://1drv.ms/v/s!AlScPmE9PLAKgRXSKm1EbEi-ryqd?e=ICMUdX

Can't remember a day without finding new bugs!

**Comment by chrissainty** [comment]
@KKacer What you've found isn't a bug, it's an error in your code. The compiler error correctly states that components can't start with a lowercase character. In your video, your component is named _test2.razor_, lower case _t_. It should be named _Test2.razor_.

**Comment by KKacer** [comment]
Oops!
Thanks @chrissainty , didn't know that I should change the filename to capital.
Thought that should be treated as a page not component, and the system wrongly think it is still a component.
BTW I heard everything in blazor will be treated as a component, and that's the case for it.
Searched the error I got, no result came up.
I really didn't list or record every bug I faced, those got me tired a bit previous week.
Even a day before encountered another bug about changing layouts which didn't take effect, closed/cleared everything, renamed the new layout to have  the real effects.
